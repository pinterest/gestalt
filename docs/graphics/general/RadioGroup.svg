<svg xmlns="http://www.w3.org/2000/svg" width="312" height="180" fill="none"><circle cx="102" cy="62" r="12" fill="#005C62"/><rect width="8" height="8" x="98" y="58" fill="#fff" rx="4"/><path fill="#fff" fill-rule="evenodd" d="M90 102c0-6.627 5.373-12 12-12s12 5.373 12 12-5.373 12-12 12-12-5.373-12-12Z" clip-rule="evenodd"/><path fill="#009990" fill-rule="evenodd" d="M92 102c0 5.523 4.477 10 10 10s10-4.477 10-10-4.477-10-10-10-10 4.477-10 10Zm10-12c-6.627 0-12 5.373-12 12s5.373 12 12 12 12-5.373 12-12-5.373-12-12-12Z" clip-rule="evenodd"/><path fill="#fff" fill-rule="evenodd" d="M90 142c0-6.627 5.373-12 12-12s12 5.373 12 12-5.373 12-12 12-12-5.373-12-12Z" clip-rule="evenodd"/><path fill="#009990" fill-rule="evenodd" d="M92 142c0 5.523 4.477 10 10 10s10-4.477 10-10-4.477-10-10-10-10 4.477-10 10Zm10-12c-6.627 0-12 5.373-12 12s5.373 12 12 12 12-5.373 12-12-5.373-12-12-12Z" clip-rule="evenodd"/><rect width="100" height="8" x="122" y="138" fill="#009990" fill-opacity=".5" rx="4"/><rect width="100" height="8" x="122" y="98" fill="#009990" fill-opacity=".5" rx="4"/><rect width="100" height="8" x="122" y="58" fill="#009990" fill-opacity=".5" rx="4"/><rect width="100" height="8" x="90" y="26" fill="#009990" fill-opacity=".5" rx="4"/></svg>