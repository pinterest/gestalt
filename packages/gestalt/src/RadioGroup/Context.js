// @flow strict
import { createContext, useContext, type Context } from 'react';

type RadioGroupContextType = {|
  parentName: string,
|};

const initialContextState = {
  parentName: '',
};

const context: Context<RadioGroupContextType> =
  createContext<RadioGroupContextType>(initialContextState);

const RadioGroupContextProvider = context.Provider;

function useRadioGroupContext(): RadioGroupContextType {
  const contextContent = useContext(context);
  return contextContent;
}

export { RadioGroupContextProvider, useRadioGroupContext };
