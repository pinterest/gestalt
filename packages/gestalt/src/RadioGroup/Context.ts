import { Context, createContext, useContext } from 'react';

type RadioGroupContextType = {
  parentName: string;
};

const initialContextState = {
  parentName: '',
} as const;

const context: Context<RadioGroupContextType> =
  createContext<RadioGroupContextType>(initialContextState);

const RadioGroupContextProvider = context.Provider;

function useRadioGroupContext(): RadioGroupContextType {
  const contextContent = useContext(context);
  return contextContent;
}

export { RadioGroupContextProvider, useRadioGroupContext };
