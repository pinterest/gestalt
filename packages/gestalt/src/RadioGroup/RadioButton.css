.container {
  align-items: flex-start;
  display: flex;
  justify-content: flex-start;
}

.wrapper .container:not(:first-child) {
  margin-top: var(--sema-space-300);
}

/* outer container */

.outerContainer {
  border-radius: var(--sema-rounding-circle);
  position: relative;
}

html[dir="rtl"] .outerContainer {
  margin-left: var(--sema-space-200);
}

html:not([dir="rtl"]) .outerContainer {
  margin-right: var(--sema-space-200);
}

.outerContainerSm {
  box-sizing: border-box;
  margin-top: var(--sema-space-50);
  min-height: 16px;
  min-width: 16px;
}

.outerContainerMd {
  box-sizing: border-box;
  min-height: 24px;
  min-width: 24px;
}

.focusedOutline {
  outline: 2px solid var(--sema-color-border-focus-outer-default);
}

.focusedUncheckedBorderVisible {
  border-width: 0;
}

.focusedUncheckedBorder {
  border: 1px solid var(--sema-color-border-interactive);
}

.focusedUncheckedBorder:hover {
  border: 1.5px solid var(--sema-color-hover-border-interactive);
}

.focusedUncheckedBorder:active {
  border: 1.5px solid var(--sema-color-pressed-border-interactive);
}

.focusedCheckedBorder {
  border: 2px solid var(--sema-color-border-focus-inner-default);
}

.disabledBorder,
.checkedBorder {
  border-width: 0;
}

.uncheckedBorder {
  border: 1px solid var(--sema-color-border-interactive);
}

.uncheckedBorderHovered {
  border: 1.5px solid var(--sema-color-hover-border-interactive);
}

.uncheckedBorderActive {
  border: 1.5px solid var(--sema-color-pressed-border-interactive);
}

/* radio */

.radio {
  border-radius: var(--sema-rounding-circle);
  box-sizing: border-box;
  height: 100%;
  position: absolute;
  width: 100%;
}

.uncheckedRadio {
  border-width: 0;
}

.checkedRadio {
  border-color: var(--sema-color-background-selected);
}

.disabledRadio {
  border-color: var(--sema-color-background-disabled);
}

.disabledUncheckedRadio {
  background-color: var(--sema-color-background-disabled);
}

.borderRadioSm {
  border-style: solid;
  border-width: 6px;
}

.borderRadioMd {
  border-style: solid;
  border-width: 8px;
}

.focusedBorderRadioSm {
  border-style: solid;
  border-width: 4px;
}

.focusedBorderRadioMd {
  border-style: solid;
  border-width: 6px;
}

.checkedTransitionAnimation {
  transition: border var(--comp-radiogroupbutton-motion-border-scale-duration)
    var(--comp-radiogroupbutton-motion-border-scale-easing)
    var(--comp-radiogroupbutton-motion-border-scale-duration);
}

.uncheckedTransitionAnimation {
  transition-duration: 150ms;
  transition-property: opacity;
  transition-timing-function: var(
    --comp-radiogroupbutton-motion-border-fade-easing
  );
}

.checkedOpacityTransitionAnimation {
  opacity: var(--opacity-0);
}

.noTransitionAnimation {
  transition: none;
}

.noTransitionAnimationDelay {
  transition-delay: var(--sema-motion-opacity-duration-instant);
}

/* input */

.input {
  appearance: none;
  box-sizing: border-box;
  margin: 0;
  opacity: var(--sema-opacity-0);
  outline: 0;
  position: absolute;
}

.inputSm {
  height: 16px;
  width: 16px;
}

.inputMd {
  height: 24px;
  width: 24px;
}
