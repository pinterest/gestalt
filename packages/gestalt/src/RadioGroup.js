// @flow strict
import { type Node, useEffect } from 'react';
import { useGlobalEventsHandler } from 'gestalt';
import Fieldset from './Fieldset.js';
import Flex from './Flex.js';
import { RadioGroupContextProvider } from './RadioGroup/Context.js';
import RadioGroupButton from './RadioGroupButton.js';

type Props = {
  /**
   * A collection of RadioGroup.RadioButtons representing the available options, as well as any Labels or layout components (Box, Flex, etc.), if needed. Other components such as Checkboxes should not be included. Note that children can be grouped into organizational components if desired.
   *
   */
  children: Node,
  /**
   * A unique identifier for this RadioGroup.
   *
   */
  id: string,
  /**
   * The description of the radio group that tells users what is being asked of them.
   *
   */
  legend: string,
  /**
   * Determines the layout of the group. See the [direction](https://gestalt.pinterest.systems/web/radiogroup#Direction) variant to learn more.
   *
   */
  direction?: 'column' | 'row',
  /**
   * Adds an error message below the group of radio buttons. See the [error](https://gestalt.pinterest.systems/web/radiogroup#With-an-error) variant to learn more.
   *
   */
  errorMessage?: string,
  /**
   * Whether the legend should be visible or not. If hidden, the legend is still available for screen reader users, but does not appear visually. See the [legend visibility](https://gestalt.pinterest.systems/web/radiogroup#Legend-visibility) variant to learn more.
   *
   */
  legendDisplay?: 'visible' | 'hidden',
};

/**
 *  [RadioGroups](https://gestalt.pinterest.systems/web/radiogroup) are used for selecting only 1 item from a list of 2 or more items. If you need multiple selection or have only one option, use [Checkbox](https://gestalt.pinterest.systems/web/checkbox). If you need to provide a binary on/off choice that takes effect immediately, use [Switch](https://gestalt.pinterest.systems/web/switch).
 *
 */
function RadioGroup({
  children,
  direction = 'column',
  errorMessage,
  id,
  legend,
  legendDisplay = 'visible',
}: Props): Node {
  // Consume GlobalEventsHandlerProvider
  const { radioGroupHandlers } = useGlobalEventsHandler() ?? {
    radioGroupHandlers: undefined,
  };

  useEffect(() => {
    if (radioGroupHandlers?.onRender) radioGroupHandlers?.onRender();
  }, [radioGroupHandlers]);

  return (
    <RadioGroupContextProvider value={{ parentName: 'RadioGroup' }}>
      <Fieldset id={id} legend={legend} errorMessage={errorMessage} legendDisplay={legendDisplay}>
        <Flex
          direction={direction}
          gap={direction === 'row' ? { row: 4, column: 0 } : { column: 2, row: 0 }}
        >
          {children}
        </Flex>
      </Fieldset>
    </RadioGroupContextProvider>
  );
}

RadioGroup.displayName = 'RadioGroup';

RadioGroup.RadioButton = RadioGroupButton;

export default RadioGroup;
