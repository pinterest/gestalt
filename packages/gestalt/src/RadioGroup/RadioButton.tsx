import { forwardRef, ReactNode } from 'react';
import classnames from 'classnames';
import { useRadioGroupContext } from './Context';
import styles from './RadioButton.css';
import Badge from '../Badge';
import Box from '../Box';
import Flex from '../Flex';
import Label from '../Label';
import FormHelperText from '../sharedSubcomponents/FormHelperText';
import Text from '../Text';
import useFocusVisible from '../useFocusVisible';
import useInteractiveStates from '../utils/useInteractiveStates';
import useTapScaleAnimation from '../utils/useTapScaleAnimation';

type BadgeType = {
  text: string;
  type?:
    | 'info'
    | 'error'
    | 'warning'
    | 'success'
    | 'neutral'
    | 'recommendation'
    | 'darkWash'
    | 'lightWash';
};

type Props = {
  badge?: BadgeType;
  checked?: boolean;
  disabled?: boolean;
  id: string;
  image?: ReactNode;
  label?: string;
  name?: string;
  onChange: (arg1: { event: React.ChangeEvent<HTMLInputElement>; checked: boolean }) => void;
  ref?: HTMLInputElement; // eslint-disable-line react/no-unused-prop-types,
  size?: 'sm' | 'md';
  helperText?: string;
  value: string;
};

const RadioGroupButtonWithForwardRef = forwardRef<HTMLInputElement, Props>(function RadioButton(
  { checked, disabled, id, image, label, name, onChange, helperText, value, badge, size }: Props,
  ref,
) {
  const { isFocusVisible } = useFocusVisible();

  const tapScaleAnimation = useTapScaleAnimation();

  const {
    handleOnMouseEnter,
    handleOnMouseLeave,
    handleOnBlur,
    handleOnFocus,
    handleOnMouseDown,
    handleOnMouseUp,
    isFocused,
    isHovered,
    isActive: isPressed,
  } = useInteractiveStates();

  const { parentName } = useRadioGroupContext();

  if (parentName !== 'RadioGroup') {
    throw new Error(
      `RadioGroup.RadioButton must be used within a [RadioGroup](https://gestalt.pinterest.systems/web/radiogroup).`,
    );
  }

  return (
    <div className={classnames(styles.container)}>
      {/* OUTER CONTAINER */}
      {/* eslint-disable-next-line jsx-a11y/no-static-element-interactions */}
      <div
        ref={tapScaleAnimation.elementRef}
        className={classnames(styles.outerContainer, tapScaleAnimation.classes, {
          [styles.outerContainerSm]: size === 'sm',
          [styles.outerContainerMd]: size === 'md',

          // disabled
          [styles.disabledBorder]: disabled,

          // outer/inner focus with keyboard navigation
          [styles.focusedOutline]: !disabled && isFocused && isFocusVisible,
          [styles.focusedCheckedBorder]: !disabled && checked && isFocused && isFocusVisible,
          // we remove this border when the RADIO element diplays as checked
          [styles.checkedBorder]: !disabled && checked && !isFocused,

          // unchecked states
          [styles.uncheckedBorder]: !disabled && !checked && !isFocused,
          [styles.uncheckedBorderHovered]: !disabled && !checked && isHovered && !isFocused,
          [styles.uncheckedBorderActive]: !disabled && !checked && isPressed,

          // We remove the border when we keyboard navigate into the RadioGroup for the first time
          [styles.focusedUncheckedBorderVisible]:
            !disabled && !checked && isFocused && isFocusVisible,
          // this supports the edge case when we click on input, set focus, but we don't pass the check true value. It's an implementation mistake but it would break the UI
          [styles.focusedUncheckedBorder]: !disabled && !checked && isFocused && !isFocusVisible,
        })}
        onMouseDown={() => tapScaleAnimation.handleMouseDown()}
        onMouseUp={() => tapScaleAnimation.handleMouseUp()}
      >
        {/* RADIO */}
        <div
          className={classnames(styles.radio, {
            [styles.checkedTransitionAnimation]: checked,
            [styles.noTransitionAnimationDelay]: checked,

            [styles.borderRadioSm]: checked && size === 'sm',
            [styles.borderRadioMd]: checked && size === 'md',

            [styles.disabledUncheckedRadio]: disabled && !checked,
            [styles.disabledRadio]: disabled && checked,

            [styles.uncheckedRadio]: !disabled && !checked,
            [styles.checkedRadio]: !disabled && checked,

            // we decrease the border width when it's displayed along the inner white border to keep the same size for the inner hole
            [styles.focusedBorderRadioSm]:
              !disabled && checked && size === 'sm' && isFocused && isFocusVisible,
            [styles.focusedBorderRadioMd]:
              !disabled && checked && size === 'md' && isFocused && isFocusVisible,
          })}
        />
        {/* INPUT */}
        <input
          ref={ref}
          checked={checked}
          className={classnames(styles.input, {
            [styles.disabledInput]: disabled,
            [styles.inputSm]: size === 'sm',
            [styles.inputMd]: size === 'md',
          })}
          disabled={disabled}
          id={id}
          name={name}
          onBlur={handleOnBlur}
          onChange={(event) => onChange({ checked: event.target.checked, event })}
          onFocus={handleOnFocus}
          onMouseDown={handleOnMouseDown}
          onMouseEnter={handleOnMouseEnter}
          onMouseLeave={handleOnMouseLeave}
          onMouseUp={handleOnMouseUp}
          type="radio"
          value={value}
        />
      </div>
      {Boolean(image) && <Box marginEnd={2}>{image}</Box>}
      <Flex direction="column">
        <Flex direction="row" gap={1}>
          {label && (
            <Label htmlFor={id}>
              <Box marginTop={size === 'sm' ? 0 : 0.25}>
                <Text color={disabled ? 'subtle' : undefined} size={size === 'sm' ? '200' : '300'}>
                  {label}
                </Text>
              </Box>
            </Label>
          )}
          {badge && (
            <Flex.Item alignSelf="end" minWidth={0}>
              <Box dangerouslySetInlineStyle={{ __style: { top: '1px' } }} position="relative">
                <Badge position="top" text={badge.text} type={badge.type || 'info'} />
              </Box>
            </Flex.Item>
          )}
        </Flex>
        {label && helperText ? <FormHelperText id={`${id}-helperText`} text={helperText} /> : null}
      </Flex>
    </div>
  );
});

RadioGroupButtonWithForwardRef.displayName = 'RadioGroup.RadioButton';

export default RadioGroupButtonWithForwardRef;
