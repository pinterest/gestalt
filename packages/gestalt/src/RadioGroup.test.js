// @flow strict
import { create } from 'react-test-renderer';
import RadioGroup from './RadioGroup';

describe('RadioGroup', () => {
  it('renders', () => {
    const tree = create(
      <RadioGroup id="testing-example" legend="testing">
        <RadioGroup.RadioButton id="choice-1" name="choice" onChange={() => {}} value="choice-1" />
      </RadioGroup>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a hidden legend', () => {
    const tree = create(
      <RadioGroup id="hiddenEx" legend="testing" legendDisplay="hidden">
        <RadioGroup.RadioButton id="choice-1" name="choice" onChange={() => {}} value="choice-1" />
      </RadioGroup>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders in different direction', () => {
    const tree = create(
      <RadioGroup direction="row" id="hiddenEx" legend="testing" legendDisplay="hidden">
        <RadioGroup.RadioButton id="choice-1" name="choice" onChange={() => {}} value="choice-1" />
      </RadioGroup>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders an error', () => {
    const tree = create(
      <RadioGroup errorMessage="please pick one" id="testingErrorEx" legend="testing">
        <RadioGroup.RadioButton id="choice-1" name="choice" onChange={() => {}} value="choice-1" />
      </RadioGroup>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
