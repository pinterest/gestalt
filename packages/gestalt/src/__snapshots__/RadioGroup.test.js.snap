// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup renders 1`] = `
<fieldset
  className="unstyled p0 m0"
>
  <legend
    className="label formLabel paddingX0"
  >
    <div
      className="Text fontSize100 defaultText alignStart breakWord fontWeightNormal"
    >
      testing
    </div>
  </legend>
  <div
    className="Flex rowGap0 columnGap2 xsDirectionColumn"
  >
    <div
      className="FlexItem"
    >
      <div
        className="box marginEndN1 marginStartN1 xsDisplayFlex xsItemsStart"
      >
        <div
          className="box paddingX1"
        >
          <div
            className="BgEnabled Border BorderUnchecked sizeMd RadioButton"
          >
            <input
              checked={false}
              className="input sizeMd InputEnabled"
              disabled={false}
              id="choice-1"
              name="choice"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              type="radio"
              value="choice-1"
            />
          </div>
        </div>
        <div
          className="Flex rowGap0 columnGap0 xsDirectionColumn"
        />
      </div>
    </div>
  </div>
</fieldset>
`;

exports[`RadioGroup renders a hidden legend 1`] = `
<fieldset
  className="unstyled p0 m0"
>
  <legend
    className="label formLabel paddingX0 visuallyHidden"
  >
    <div
      className="Text fontSize100 defaultText alignStart breakWord fontWeightNormal"
    >
      testing
    </div>
  </legend>
  <div
    className="Flex rowGap0 columnGap2 xsDirectionColumn"
  >
    <div
      className="FlexItem"
    >
      <div
        className="box marginEndN1 marginStartN1 xsDisplayFlex xsItemsStart"
      >
        <div
          className="box paddingX1"
        >
          <div
            className="BgEnabled Border BorderUnchecked sizeMd RadioButton"
          >
            <input
              checked={false}
              className="input sizeMd InputEnabled"
              disabled={false}
              id="choice-1"
              name="choice"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              type="radio"
              value="choice-1"
            />
          </div>
        </div>
        <div
          className="Flex rowGap0 columnGap0 xsDirectionColumn"
        />
      </div>
    </div>
  </div>
</fieldset>
`;

exports[`RadioGroup renders an error 1`] = `
<fieldset
  aria-describedby="testingErrorEx-error"
  className="unstyled p0 m0"
>
  <legend
    className="label formLabel paddingX0"
  >
    <div
      className="Text fontSize100 defaultText alignStart breakWord fontWeightNormal"
    >
      testing
    </div>
  </legend>
  <div
    className="Flex rowGap0 columnGap2 xsDirectionColumn"
  >
    <div
      className="FlexItem"
    >
      <div
        className="box marginEndN1 marginStartN1 xsDisplayFlex xsItemsStart"
      >
        <div
          className="box paddingX1"
        >
          <div
            className="BgEnabled Border BorderUnchecked sizeMd RadioButton"
          >
            <input
              checked={false}
              className="input sizeMd InputEnabled"
              disabled={false}
              id="choice-1"
              name="choice"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              type="radio"
              value="choice-1"
            />
          </div>
        </div>
        <div
          className="Flex rowGap0 columnGap0 xsDirectionColumn"
        />
      </div>
    </div>
  </div>
  <div
    className="box marginTop2"
  >
    <div
      className="Text fontSize100 errorText alignStart breakWord fontWeightNormal"
    >
      <span
        className="formErrorMessage"
        id="testingErrorEx-error"
      >
        <div
          className="box"
          role="alert"
        >
          <div
            className="Flex rowGap2 columnGap2 xsDirectionRow xsItemsCenter"
          >
            <div
              className="FlexItem"
            >
              <svg
                aria-hidden={true}
                aria-label=""
                className="icon errorIcon iconBlock"
                height={16}
                role="img"
                viewBox="0 0 24 24"
                width={16}
              >
                <path
                  d="test-file-stub"
                />
              </svg>
            </div>
            <div
              className="FlexItem"
            >
              please pick one
            </div>
          </div>
        </div>
      </span>
    </div>
  </div>
</fieldset>
`;

exports[`RadioGroup renders in different direction 1`] = `
<fieldset
  className="unstyled p0 m0"
>
  <legend
    className="label formLabel paddingX0 visuallyHidden"
  >
    <div
      className="Text fontSize100 defaultText alignStart breakWord fontWeightNormal"
    >
      testing
    </div>
  </legend>
  <div
    className="Flex rowGap4 columnGap0 xsDirectionRow"
  >
    <div
      className="FlexItem"
    >
      <div
        className="box marginEndN1 marginStartN1 xsDisplayFlex xsItemsStart"
      >
        <div
          className="box paddingX1"
        >
          <div
            className="BgEnabled Border BorderUnchecked sizeMd RadioButton"
          >
            <input
              checked={false}
              className="input sizeMd InputEnabled"
              disabled={false}
              id="choice-1"
              name="choice"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              type="radio"
              value="choice-1"
            />
          </div>
        </div>
        <div
          className="Flex rowGap0 columnGap0 xsDirectionColumn"
        />
      </div>
    </div>
  </div>
</fieldset>
`;
